util/buildinfo/buildinfo.go

```diff
@@ -197,6 +211,25 @@ func decodeDeps(key string, attrs map[string]*string) (map[string]binfotypes.Bui
 	return res, nil
 }
 
+// dedupSources deduplicates main sources from dependencies ones.
+// FIXME: if a main source is also used in a dependency build, it will be removed. We should identify this case.
+func dedupSources(bi binfotypes.BuildInfo) []binfotypes.Source {
```

Comment:
This needs to be recursive function as described in https://github.com/moby/buildkit/pull/2684#discussion_r818147511